File: clients/javascript/model/player-view.js
const Player = require.main.require('./model/player');
const Quest = require.main.require('./model/quest');
/**
 * Current game's state
 */
class PlayerView {
    /**
     * Current tick number
     */
    currentTick;
    /**
     * TODO - Document
     */
    you;
    /**
     * TODO - Document
     */
    other;
    /**
     * TODO - Document
     */
    quests;

    constructor(currentTick, you, other, quests) {
        this.currentTick = currentTick;
        this.you = you;
        this.other = other;
        this.quests = quests;
    }

    /**
     * Read PlayerView from input stream
     */
    static async readFrom(stream) {
        let currentTick;
        currentTick = await stream.readInt();
        let you;
        you = await Player.readFrom(stream);
        let other;
        other = [];
        for (let otherCount = await stream.readInt(); otherCount > 0; otherCount--) {
            let otherElement;
            otherElement = await Player.readFrom(stream);
            other.push(otherElement);
        }
        let quests;
        quests = [];
        for (let questsCount = await stream.readInt(); questsCount > 0; questsCount--) {
            let questsElement;
            questsElement = await Quest.readFrom(stream);
            quests.push(questsElement);
        }
        return new PlayerView(currentTick, you, other, quests);
    }

    /**
     * Write PlayerView to output stream
     */
    async writeTo(stream) {
        let currentTick = this.currentTick;
        await stream.writeInt(currentTick);
        let you = this.you;
        await you.writeTo(stream);
        let other = this.other;
        await stream.writeInt(other.length);
        for (let otherElement of other) {
            await otherElement.writeTo(stream);
        }
        let quests = this.quests;
        await stream.writeInt(quests.length);
        for (let questsElement of quests) {
            await questsElement.writeTo(stream);
        }
    }
}
module.exports = PlayerView

File: clients/python/model/player_view.py
from model.player import Player
from model.quest import Quest
from stream_wrapper import StreamWrapper
from typing import List

class PlayerView:
    """Current game's state"""

    __slots__ = ("current_tick","you","other","quests",)

    current_tick: int
    you: Player
    other: List[Player]
    quests: List[Quest]

    def __init__(self, current_tick: int, you: Player, other: List[Player], quests: List[Quest]):
        self.current_tick = current_tick
        """Current tick number"""
        self.you = you
        """TODO - Document"""
        self.other = other
        """TODO - Document"""
        self.quests = quests
        """TODO - Document"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "PlayerView":
        """Read PlayerView from input stream
        """
        current_tick = stream.read_int()
        you = Player.read_from(stream)
        other = []
        for _ in range(stream.read_int()):
            other_element = Player.read_from(stream)
            other.append(other_element)
        quests = []
        for _ in range(stream.read_int()):
            quests_element = Quest.read_from(stream)
            quests.append(quests_element)
        return PlayerView(current_tick, you, other, quests)
    
    def write_to(self, stream: StreamWrapper):
        """Write PlayerView to output stream
        """
        stream.write_int(self.current_tick)
        self.you.write_to(stream)
        stream.write_int(len(self.other))
        for element in self.other:
            element.write_to(stream)
        stream.write_int(len(self.quests))
        for element in self.quests:
            element.write_to(stream)
    
    def __repr__(self):
        return "PlayerView(" + \
            repr(self.current_tick) + \
            ", " + \
            repr(self.you) + \
            ", " + \
            repr(self.other) + \
            ", " + \
            repr(self.quests) + \
            ")"